added cloud build event type

capri-xiyue opened the pull request:
added cloud build event  and data proto

capri-xiyue commented:
assign @yolocs

jskeet commented on proto/google/events/cloud/build/v1/data.proto:
```diff
@@ -0,0 +1,631 @@
+    EXPIRED = 9;
+  }
+
+  // Output only. Unique identifier of the build.
```
Drive-by review as I stop for the night - it's worth going through all the docs thinking about this as "this is event data; it's only ever consumed". So "output only" and defaults don't really mean very much in that context.

But we can tweak all of that later, of course.

yolocs replied:
+1. I don't think we need the OUTPUT_ONLY option.

yolocs commented on proto/google/events/cloud/build/v1/data.proto:
```diff
@@ -0,0 +1,631 @@
+
+syntax = "proto3";
+
+package google.events.cloud.build.v1;
+
+option csharp_namespace = "Google.Events.Protobuf.Cloud.Build.V1";
+import "google/api/field_behavior.proto";
+import "google/protobuf/duration.proto";
+import "google/protobuf/empty.proto";
+import "google/protobuf/timestamp.proto";
```
nit: put csharp_namespace after the imports to be consistent with others.

yolocs commented on proto/google/events/cloud/build/v1/events.proto:
```diff
@@ -0,0 +1,31 @@
+
+// The CloudEvent raised when your build's state changes.
+message CloudBuildEvent {
+  option (google.events.cloud_event_type) =
+      "google.cloud.build.v1.event";
```
Do we only have one event type? In general the type should look like something `google.cloud.build.something.v1.happened`

capri-xiyue replied:
Yep. Currently we only have one event type for cloudbuildsource

yolocs commented on third_party/googleapis/google/api/README.md:
```diff
@@ -0,0 +1,5 @@
+This folder contains the schema of the configuration model for the API services
```
@jskeet I'm not sure if we have some process to follow to import third party protos. If so, let us know.

jskeet replied:
I would hope we wouldn't need anything from `google/api`. The only import is for field_behavior.proto, which we shouldn't use. I'm hoping this PR can turn into three files - data.proto and events.proto as new files, and an update to the README.

capri-xiyue replied:
@jskeet If we don't import `field_behavior.proto`, the build will fail as follows: https://travis-ci.org/github/googleapis/google-cloudevents/builds/714923575
Is that expected?

jskeet replied:
That failure shows you still trying to import the proto. You should:

- Remove the proto
- Remove the import
- Remove all uses of the annotation

Basically it's an annotation that only makes sense in an API proto, which this isn't.

yolocs commented on .idea/workspace.xml:
```diff
@@ -0,0 +1,53 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Remove this folder please.

yolocs commented:
@hrasadi

capri-xiyue commented:
@nachocano

hrasadi commented:
Overall LGTM. In terms of the event type string, is there a way for us to double check with the the service owners?

capri-xiyue commented:
> Overall LGTM. In terms of the event type string, is there a way for us to double check with the the service owners?

@hrasadi When you say `event type string`, do you mean this one  `google.cloud.build.status.v1.changed`? What do you mean by `double check with the the service owners`? I think I'm gonna do some change for this file https://github.com/google/knative-gcp/blob/b3d705927b0842a190661df14936aefdd22942ad/config/core/resources/cloudbuildsource.yaml#L25 to make it match? Is that what you mean for the `double check`?

hrasadi commented:
I mean if possible, asking the Cloud Build team what the event type should be called.

capri-xiyue commented:
> I mean if possible, asking the Cloud Build team what the event type should be called.

I refer to this https://cloud.google.com/cloud-build/docs/subscribe-build-notifications to define the naming of event type. 
`Cloud Build publishes messages on a Google Pub/Sub topic when your build's state changes, such as when your build is created, when your build transitions to a working state, and when your build completes.`
I will aslo consult Cloud Build team to see whether they have any insights.

yolocs commented:
FWIW, the actual cloud build resource is called `build`. According to the current convention, the event type should be something like `google.cloud.build.build.v1.statusChanged`. But the double "build" might cause the illusion it's a mistake.

We should consult cloud build team for sure.

@jskeet @hrasadi any thoughts?

jskeet commented:
Hmm. I *think* I prefer the odd-looking `google.cloud.build.build.v1.statusChanged` event type. I think the consistency of that will help with consistency in others aspects which *don't* have the oddity, whereas if we decide to break the convention for this aspect of Cloud Build, it might make some other aspects less conventional as a side-effect. But I'll readily admit I haven't put much time/thought into it.

capri-xiyue commented:
@jskeet @hrasadi @yolocs 
I just got reply from Cloud Build Team, 
`I would have expected "google.cloudbuild.v1.status.changed" or possibly "google.cloudbuild.v1.build.status.changed"?
Meaning effectively (in my head) that the service is cloudbuild.v1, the object in question is a build, the field in question is build.status, and the event is that the status changed.
But as usual, naming is hard...`
But what Cloud Build eam suggests doesn't follow our convention.
For me I prefer `google.cloud.build.build.v1.statusChanged` since it follows our convention.

jskeet commented:
Could I suggest that we take this discussion internal at this point? I think it'll probably be simpler.

capri-xiyue commented:
> Could I suggest that we take this discussion internal at this point? I think it'll probably be simpler.

sure

jskeet commented on proto/google/events/cloud/build/v1/data.proto:
```diff
@@ -0,0 +1,626 @@
+// Build event data
+// Common build format for Google Cloud Platform API operations.
+// Copied from
+// https://github.com/googleapis/googleapis/blob/master/google/devtools/cloudbuild/v1/cloudbuild.proto,
```
Nit: why the comma at the end? Was there meant to be another line with additional information?

jskeet commented on proto/google/events/cloud/build/v1/data.proto:
```diff
@@ -0,0 +1,626 @@
+  // granularity. If this amount of time elapses, work on the build will cease
+  // and the build status will be `TIMEOUT`.
+  //
+  // Default time is ten minutes.
```
Will this always actually be populated, i.e. if the user didn't specify it, it will be present with a value of 10 minutes? If so, let's remove the "default time" part.

capri-xiyue replied:
Yep. If user doesn't specify it, it will be present with a value of 10 minutes.

jskeet commented on proto/google/events/cloud/build/v1/data.proto:
```diff
@@ -0,0 +1,626 @@
+
+  // Google Cloud Storage object containing the source.
+  //
+  // This object must be a gzipped archive file (`.tar.gz`) containing source to
```
Remove this sentence, given that the event is providing information rather than making requirements of callers?

jskeet commented on proto/google/events/cloud/build/v1/data.proto:
```diff
@@ -0,0 +1,626 @@
+
+// Location of the source in a Google Cloud Source Repository.
+message RepoSource {
+  // ID of the project that owns the Cloud Source Repository. If omitted, the
```
As with the earlier timeout field: will this always be populated?

jskeet commented on proto/google/events/cloud/build/v1/data.proto:
```diff
@@ -0,0 +1,626 @@
+  // project ID requesting the build is assumed.
+  string project_id = 1;
+
+  // Required. Name of the Cloud Source Repository.
```
Please remove the Required part.

jskeet commented on proto/google/events/cloud/build/v1/data.proto:
```diff
@@ -0,0 +1,626 @@
+  // The location of the source files to build.
+  Source source = 3;
+
+  // Required. The operations to be performed on the workspace.
```
Please remove the "Required." part

jskeet commented on proto/google/events/cloud/build/v1/data.proto:
```diff
@@ -0,0 +1,626 @@
+
+// A step in the build pipeline.
+message BuildStep {
+  // Required. The name of the container image that will run this particular
```
Please remove the required part.

jskeet commented on proto/google/events/cloud/build/v1/data.proto:
```diff
@@ -0,0 +1,626 @@
+  // Option to specify a `WorkerPool` for the build.
+  // Format: projects/{project}/locations/{location}/workerPools/{workerPool}
+  //
+  // This field is experimental.
```
Remove "experimental" part? (Not that it should be in the original API either, admittedly... stable APIs should be stable...)
